_posts/2017-06-15-REST-api-NodeJS-koa.md

```diff
@@ -0,0 +1,230 @@
+
+[You can find the code on Github.](https://github.com/ivarvh/movielistr-backend-ts-ioc)
+
+# Setup
+
+Setting up a node project with TypeScript doesn’t require a lot of effort, the following commands are enough to get started.
+
+{% highlight shell %}
+mkdir <project-name>
+cd <project-name>
+mkdir src
+npm init //follow the setup
+npm install --save-dev TypeScript tsc //install TypeScript and the TypeScript compiler
+tsc --init //generates a tsconfig.json, a config for the TypeScript compiler
+{% endhighlight %}
+
+This is it, you still have to tinker with the tsconfig.json to get it to your liking, but after that you can just start writing code.
+
+# Using async / await
+
+I want to start with talking about the async / await features. They were what really made this code so fast to write and easy to read. The async keyword marks a function that will always return a promise. The await keyword will automatically unwrap the value from the promise and continue the code when the promise has been resolved. A small example:
+
+{% highlight coffeescript %}  
+    const promiseFn = (): Promise<string> => {
+	    return Promise.resolve("Hello World");
+    }
+ 
+    // Old way:
+ 
+    const asyncFn = () => {
+	    promiseFn()
+		    .then(value => {
+			    console.log(value);
+            });
+    }
+ 
+    // With async await
+ 
+    const asyncFn = async () => {
+	    const text = await promiseFn();
+	    console.log(text);
+    }
+{% endhighlight %}
+
+You can see how readable it is with the async await syntax. You can write asynchronous code in a synchronous way and I used it heavily everywhere in my code. I think this is one the things that will really make writing JavaScript fun. No more callbacks, no more boilerplate code, just the important bits. For error handling you can rely on try catch statements to catch errors and act on them.
+ 
+To use the async / await syntax, you can have to add "esnext.asynciterable" to the lib array in the tsconfig.json file.
+
+# The libraries I used
+
+### Koa
+
+Koa is a small node library to create REST APIs. It was made by the guys who created express. It takes advantage of the new ES6 feature of generator functions and it allows you to write very readable code by using the async/await features (that are based on the generator functions). For a full understanding of koa and generator functions, I suggest the [koa course](https://www.pluralsight.com/courses/JavaScript-koa-introduction) on pluralsight from Hammarberg. 
+ 
+Koa relies heavily on middleware, so for every "step" of the process we need middleware. For instance koa-bodyparser middleware will parse the request body to json, the koa-logger middleware will log all the incoming requests and the koa-router middleware will make it easy for us to configure the url mapping to certain actions. These middlewares are installed apart from the koa framework or you can write them yourself.
```

Comment:
These "koa-xxx"-parts are dependencies used in the code? If so, you could wrap them in backticks.